File: algorithm/java/src/BOJ1157.java
// problem : https://www.acmicpc.net/problem/1157

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Scanner;

public class BOJ1157 {
    static int[] alphabet;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String str = br.readLine();
        str = str.toUpperCase();
        alphabet = new int[26];
        for (int i = 0, size = str.length(); i < size; i++) {
            alphabet[str.charAt(i) - 'A']++;
        }
        // System.out.println(Arrays.toString(alphabet));

        int maxValue = Integer.MIN_VALUE;
        int maxIndex = -1;

        for (int i = 0; i < alphabet.length; i++) {
            if (maxValue < alphabet[i]) {
                maxValue = alphabet[i];
                maxIndex = i;
            }
        }

        char result = (char)('A' + maxIndex);

        for (int i = 0; i < alphabet.length; i++) {
            if (i != maxIndex && maxValue <= alphabet[i]) {
                result = '?';
                break;
            }
        }

        System.out.println(result);
    }
}

File: algorithm/typescript/BOJ1157.ts
// problem : https://www.acmicpc.net/problem/1157

import fs = require("fs");

const input: string = fs.readFileSync("/dev/stdin").toString().toUpperCase();
const alphabetCnt: number[] = new Array(26).fill(0);

for (let i = 0, size = input.length; i < size; i++) {
  let index: number = input.charCodeAt(i);
  alphabetCnt[index - 65] += 1;
}
// console.log(alphabetCnt);

let maxCount: number = Number.MIN_SAFE_INTEGER;
let maxIndex: number = Number.MIN_SAFE_INTEGER;

for (let i = 0, size = alphabetCnt.length; i < size; i++) {
  if (maxCount < alphabetCnt[i]) {
    maxCount = alphabetCnt[i];
    maxIndex = i;
  }
}

let result: string = String.fromCodePoint(maxIndex + 65);

for (let i = 0, size = alphabetCnt.length; i < size; i++) {
  if (i != maxIndex && maxCount == alphabetCnt[i]) {
    result = "?";
    break;
  }
}

console.log(result);
